state-chain/pallets/cf-swapping/src/lib.rs

```diff
@@ -957,8 +1025,115 @@ pub mod pallet {
 				}
 			}
+		}
 
-			Ok(())
+		// Returns swaps that have not expired yet and should be scheduled for retry.
+		fn refund_any_expired_swaps(
+			swaps: Vec<Swap>,
+			current_block: BlockNumberFor<T>,
+		) -> Vec<Swap> {
+			let (swaps_to_retry, swaps_to_refund): (Vec<Swap>, Vec<Swap>) =
+				swaps.into_iter().partition(|swap| {
+					swap.refund_params
+						.as_ref()
+						.map_or(true, |params| current_block < params.refund_block.into())
+				});
+
+			for swap in swaps_to_refund {
+				let destination_address = swap.refund_params.unwrap().refund_address;
+				match T::EgressHandler::schedule_egress(
+					swap.from,
+					swap.input_amount,
+					destination_address,
+					None,
+				) {
+					Ok(ScheduledEgressDetails { egress_id, egress_amount, fee_withheld }) => {
+						Self::deposit_event(Event::<T>::RefundEgressScheduled {
+							swap_id: swap.swap_id,
+							egress_id,
+							amount: egress_amount,
+							egress_fee: fee_withheld,
+						});
+					},
+					Err(err) => {
+						Self::deposit_event(Event::<T>::RefundEgressIgnored {
+							swap_id: swap.swap_id,
+							asset: swap.from,
+							amount: swap.input_amount,
+							reason: err.into(),
+						});
+					},
+				}
+			}
+
+			swaps_to_retry
+		}
+
+		fn process_swaps_for_block(
+			current_block: BlockNumberFor<T>,
+			block_to_process: BlockNumberFor<T>,
+		) -> Result<(), ()> {
+			let mut swaps_to_execute = SwapQueue::<T>::take(block_to_process);
+
+			loop {
+				if swaps_to_execute.is_empty() {
+					return Ok(())
+				}
+
+				match Self::execute_batch(swaps_to_execute.clone()) {
+					Ok(successful_swaps) => {
+						Self::process_swap_outcomes(&successful_swaps);
+						// Nothing to do here, all swaps are processed for block
+						return Ok(())
+					},
+					Err(err) => {
+						let (successful_swaps, failed_swaps) = {
+							match err {
+								BatchExecutionError::PriceLimitHit {
+									successful_swaps,
+									failed_swaps,
+								} => (successful_swaps, failed_swaps),
+								BatchExecutionError::SwapLegFailed { asset, direction, amount } => {
+									Self::deposit_event(Event::<T>::BatchSwapFailed {
+										asset,
+										direction,
+										amount,
+									});
+									(vec![], swaps_to_execute)
+								},
+								BatchExecutionError::DispatchError { error } => {
+									log::error!(
+										"Failed to execute swap batch at block {:?}: {:?}",
+										block_to_process,
+										error
+									);
+									(vec![], swaps_to_execute)
+								},
+							}
+						};
+
+						let swaps_to_retry =
+							Self::refund_any_expired_swaps(failed_swaps, current_block);
+
+						let retry_block = current_block + SWAP_RETRY_DELAY_BLOCKS.into();
+
+						for swap in &swaps_to_retry {
+							Self::deposit_event(Event::<T>::SwapRescheduled {
+								swap_id: swap.swap_id,
+								execute_at: retry_block,
+							});
+						}
+
+						SwapQueue::<T>::insert(retry_block, swaps_to_retry);
```

Comment:
Also note that we check the expiry *after* execution, based on the current block. We should check expiry based on the scheduled block (or simply check before executing).

Reply:
> this seems a little dangerous to me.

Yeah, this wasn't intentional. Changed to `extend` to preserve any existing swaps.

Reply:
> We should check expiry based on the scheduled block (or simply check before executing).

OK, updated the code to check against the block that the swaps would be scheduled for.

Reply:
I feel like it would be simpler to just check for expired swaps *before* executing rather than leaving it up to the scheduler of the swap (see my other comment - we have the same issue when initially scheduling swaps).

Reply:
You did propose checking expiry based on the scheduled block as one of the options above, and I still feel that it is at least as good (if not better) than the other option. It is strange having to wait the next the next scheduled time just to realise that it should have been refunded earlier. We also don't want to schedule a swap and not end up executing it. Keep in mind that when something is scheduled, it will appear in the scheduled swaps subscription and it will be misleading not to execute it at that point.